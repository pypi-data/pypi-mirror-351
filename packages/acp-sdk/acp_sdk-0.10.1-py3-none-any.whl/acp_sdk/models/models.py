import asyncio
import uuid
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Literal, Optional, Union

from pydantic import AnyUrl, BaseModel, ConfigDict, Field

from acp_sdk.models.errors import ACPError, Error


class AnyModel(BaseModel):
    model_config = ConfigDict(extra="allow")


class Author(BaseModel):
    name: str
    email: str | None = None
    url: AnyUrl | None = None


class Contributor(BaseModel):
    name: str
    email: str | None = None
    url: AnyUrl | None = None


class LinkType(str, Enum):
    SOURCE_CODE = "source-code"
    CONTAINER_IMAGE = "container-image"
    HOMEPAGE = "homepage"
    DOCUMENTATION = "documentation"


class Link(BaseModel):
    type: LinkType
    url: AnyUrl


class DependencyType(str, Enum):
    AGENT = "agent"
    TOOL = "tool"
    MODEL = "model"


class Dependency(BaseModel):
    type: DependencyType
    name: str


class Capability(BaseModel):
    name: str
    description: str


class Metadata(BaseModel):
    annotations: AnyModel | None = None
    documentation: str | None = None
    license: str | None = None
    programming_language: str | None = None
    natural_languages: list[str] | None = None
    framework: str | None = None
    capabilities: list[Capability] | None = None
    domains: list[str] | None = None
    tags: list[str] | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    author: Author | None = None
    contributors: list[Contributor] | None = None
    links: list[Link] | None = None
    dependencies: list[Dependency] | None = None
    recommended_models: list[str] | None = None

    model_config = ConfigDict(extra="allow")


class MessagePart(BaseModel):
    name: Optional[str] = None
    content_type: Optional[str] = "text/plain"
    content: Optional[str] = None
    content_encoding: Optional[Literal["plain", "base64"]] = "plain"
    content_url: Optional[AnyUrl] = None

    model_config = ConfigDict(extra="allow")

    def model_post_init(self, __context: Any) -> None:
        if self.content is None and self.content_url is None:
            raise ValueError("Either content or content_url must be provided")
        if self.content is not None and self.content_url is not None:
            raise ValueError("Only one of content or content_url can be provided")


class Artifact(MessagePart):
    name: str


class Message(BaseModel):
    parts: list[MessagePart]
    created_at: datetime | None = Field(default_factory=lambda: datetime.now(timezone.utc))
    completed_at: datetime | None = Field(default_factory=lambda: datetime.now(timezone.utc))

    def __add__(self, other: "Message") -> "Message":
        if not isinstance(other, Message):
            raise TypeError(f"Cannot concatenate Message with {type(other).__name__}")
        return Message(
            parts=self.parts + other.parts,
            created_at=min(self.created_at, other.created_at) if self.created_at and other.created_at else None,
            completed_at=max(self.completed_at, other.completed_at)
            if self.completed_at and other.completed_at
            else None,
        )

    def __str__(self) -> str:
        return "".join(
            part.content for part in self.parts if part.content is not None and part.content_type == "text/plain"
        )

    def compress(self) -> "Message":
        def can_be_joined(first: MessagePart, second: MessagePart) -> bool:
            return (
                first.name is None
                and second.name is None
                and first.content_type == "text/plain"
                and second.content_type == "text/plain"
                and first.content_encoding == "plain"
                and second.content_encoding == "plain"
                and first.content_url is None
                and second.content_url is None
            )

        def join(first: MessagePart, second: MessagePart) -> MessagePart:
            return MessagePart(
                name=None,
                content_type="text/plain",
                content=first.content + second.content,
                content_encoding="plain",
                content_url=None,
            )

        parts: list[MessagePart] = []
        for part in self.parts:
            if len(parts) > 0 and can_be_joined(parts[-1], part):
                parts[-1] = join(parts[-1], part)
            else:
                parts.append(part)
        return Message(parts=parts, created_at=self.created_at, completed_at=self.completed_at)


AgentName = str
SessionId = uuid.UUID
RunId = uuid.UUID


class RunMode(str, Enum):
    SYNC = "sync"
    ASYNC = "async"
    STREAM = "stream"


class RunStatus(str, Enum):
    CREATED = "created"
    IN_PROGRESS = "in-progress"
    AWAITING = "awaiting"
    CANCELLING = "cancelling"
    CANCELLED = "cancelled"
    COMPLETED = "completed"
    FAILED = "failed"

    @property
    def is_terminal(self) -> bool:
        terminal_states = {RunStatus.COMPLETED, RunStatus.FAILED, RunStatus.CANCELLED}
        return self in terminal_states


class MessageAwaitRequest(BaseModel):
    type: Literal["message"] = "message"
    message: Message


class MessageAwaitResume(BaseModel):
    type: Literal["message"] = "message"
    message: Message


AwaitRequest = Union[MessageAwaitRequest]
AwaitResume = Union[MessageAwaitResume]


class Run(BaseModel):
    run_id: RunId = Field(default_factory=uuid.uuid4)
    agent_name: AgentName
    session_id: SessionId | None = None
    status: RunStatus = RunStatus.CREATED
    await_request: AwaitRequest | None = None
    output: list[Message] = []
    error: Error | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    finished_at: datetime | None = None

    def raise_for_status(self) -> "Run":
        match self.status:
            case RunStatus.CANCELLED:
                raise asyncio.CancelledError()
            case RunStatus.FAILED:
                raise ACPError(error=self.error)
            case _:
                return self


class MessageCreatedEvent(BaseModel):
    type: Literal["message.created"] = "message.created"
    message: Message


class MessagePartEvent(BaseModel):
    type: Literal["message.part"] = "message.part"
    part: MessagePart


class ArtifactEvent(BaseModel):
    type: Literal["message.part"] = "message.part"
    part: Artifact


class MessageCompletedEvent(BaseModel):
    type: Literal["message.completed"] = "message.completed"
    message: Message


class RunAwaitingEvent(BaseModel):
    type: Literal["run.awaiting"] = "run.awaiting"
    run: Run


class GenericEvent(BaseModel):
    type: Literal["generic"] = "generic"
    generic: AnyModel


class RunCreatedEvent(BaseModel):
    type: Literal["run.created"] = "run.created"
    run: Run


class RunInProgressEvent(BaseModel):
    type: Literal["run.in-progress"] = "run.in-progress"
    run: Run


class RunFailedEvent(BaseModel):
    type: Literal["run.failed"] = "run.failed"
    run: Run


class RunCancelledEvent(BaseModel):
    type: Literal["run.cancelled"] = "run.cancelled"
    run: Run


class RunCompletedEvent(BaseModel):
    type: Literal["run.completed"] = "run.completed"
    run: Run


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    error: Error


Event = Union[
    ErrorEvent,
    RunCreatedEvent,
    RunInProgressEvent,
    MessageCreatedEvent,
    ArtifactEvent,
    MessagePartEvent,
    MessageCompletedEvent,
    RunAwaitingEvent,
    GenericEvent,
    RunCancelledEvent,
    RunFailedEvent,
    RunCompletedEvent,
    MessagePartEvent,
]


class Agent(BaseModel):
    name: str
    description: str | None = None
    metadata: Metadata = Metadata()
