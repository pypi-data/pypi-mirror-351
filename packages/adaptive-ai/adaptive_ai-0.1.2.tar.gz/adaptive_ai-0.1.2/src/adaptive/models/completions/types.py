"""
Common type definitions used across the provider package.
"""

from typing import List, Optional, Dict, Any, Literal
from pydantic import BaseModel  # type: ignore

# Type alias for message roles
MessageRole = Literal["user", "assistant", "system", "tool"]


class Message(BaseModel):
    """
    Represents a single chat message in a conversation.
    """

    role: MessageRole
    content: str
    tool_call_id: Optional[str] = None
    function_call: Optional[Dict[str, str]] = None
    tool_calls: Optional[List[Dict[str, Any]]] = None


class ChatCompletionRequest(BaseModel):
    """
    Standard request format for chat completions.
    """

    messages: List[Message]
