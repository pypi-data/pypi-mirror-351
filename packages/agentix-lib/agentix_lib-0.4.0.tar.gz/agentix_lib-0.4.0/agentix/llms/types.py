from typing import Protocol, List, Dict, runtime_checkable

@runtime_checkable
class LLMProtocol(Protocol):
    """Protocol defining the interface that all LLM classes should implement."""
    
    async def call(self, messages: List[Dict[str, str]]) -> str:
        """
        Call the LLM with the given messages.
        
        Args:
            messages: List of message objects with 'role' and 'content' keys
            
        Returns:
            The model's response as a string
        """
        ...

# Base class for LLM implementations
class LLM:
    """Base class for all LLM implementations."""
    async def call(self, messages: List[Dict[str, str]]) -> str:
        """
        Call the LLM with the given messages.
        
        Args:
            messages: List of message objects with 'role' and 'content' keys
            
        Returns:
            The model's response as a string
        """
        raise NotImplementedError("Subclasses must implement this method")

# For strict typing with Protocol
LLMInterface = LLMProtocol 