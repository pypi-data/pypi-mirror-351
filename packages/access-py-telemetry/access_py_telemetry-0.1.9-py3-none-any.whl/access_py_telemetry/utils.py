"""
Copyright 2022 ACCESS-NRI and contributors. See the top-level COPYRIGHT file for details.
SPDX-License-Identifier: Apache-2.0
"""

from typing import Any
import yaml
from pathlib import Path
from dataclasses import dataclass, field


with open(Path(__file__).parent / "config.yaml", "r") as f:
    config = yaml.safe_load(f)


@dataclass
class TelemetryRegister:
    endpoint: str
    items: set[str] = field(default_factory=set)


def build_endpoints(
    config: dict[str, Any], parent: str | None = None
) -> list[TelemetryRegister]:
    """
    Recursively join the keys of the dictionary until we reach a list
    Returns list of tuples (path, endpoint_list)
    """
    parent = parent or ""
    results = []

    for key, val in config.items():
        if isinstance(val, dict):
            # Recursively process dictionaries and extend results
            nested_results = build_endpoints(val, f"{parent}/{key}" if parent else key)
            results.extend(nested_results)
        elif isinstance(val, list):
            # Add tuple of (path, list) when we find a list
            full_path = "/".join([parent, key]) if parent else key
            results.append(TelemetryRegister(full_path, set(val)))

    return results


ENDPOINTS = {
    register.endpoint.replace("/", "_"): register.endpoint
    for register in build_endpoints(config)
}

REGISTRIES = {
    register.endpoint.replace("/", "_"): register.items
    for register in build_endpoints(config)
}
