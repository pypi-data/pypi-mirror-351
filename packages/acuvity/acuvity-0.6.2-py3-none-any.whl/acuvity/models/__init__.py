"""Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT."""

from .alertevent import Alertevent, AlerteventTypedDict
from .analyzer import (
    Analyzer,
    AnalyzerTypedDict,
    DetectionMatchers,
    DetectionMatchersTypedDict,
)
from .analyzermodel import Analyzermodel, AnalyzermodelTypedDict
from .apierror import APIError
from .detector import Detector, DetectorTypedDict
from .elementalerror import Data, DataTypedDict, Elementalerror, ElementalerrorData
from .extraction import Extraction, ExtractionTypedDict
from .extractionrequest import Extractionrequest, ExtractionrequestTypedDict
from .extractionsummary import Extractionsummary, ExtractionsummaryTypedDict
from .latency import Latency, LatencyTypedDict
from .modality import Modality, ModalityTypedDict
from .principal import AuthType, Principal, PrincipalType, PrincipalTypedDict
from .principalapp import Principalapp, PrincipalappTypedDict
from .principaluser import Principaluser, PrincipaluserTypedDict
from .scanrequest import Anonymization, Scanrequest, ScanrequestTypedDict, Type
from .scanresponse import (
    Decision,
    Scanresponse,
    ScanresponseType,
    ScanresponseTypedDict,
)
from .security import Security, SecurityTypedDict
from .textualdetection import (
    Textualdetection,
    TextualdetectionType,
    TextualdetectionTypedDict,
)


__all__ = [
    "APIError",
    "Alertevent",
    "AlerteventTypedDict",
    "Analyzer",
    "AnalyzerTypedDict",
    "Analyzermodel",
    "AnalyzermodelTypedDict",
    "Anonymization",
    "AuthType",
    "Data",
    "DataTypedDict",
    "Decision",
    "DetectionMatchers",
    "DetectionMatchersTypedDict",
    "Detector",
    "DetectorTypedDict",
    "Elementalerror",
    "ElementalerrorData",
    "Extraction",
    "ExtractionTypedDict",
    "Extractionrequest",
    "ExtractionrequestTypedDict",
    "Extractionsummary",
    "ExtractionsummaryTypedDict",
    "Latency",
    "LatencyTypedDict",
    "Modality",
    "ModalityTypedDict",
    "Principal",
    "PrincipalType",
    "PrincipalTypedDict",
    "Principalapp",
    "PrincipalappTypedDict",
    "Principaluser",
    "PrincipaluserTypedDict",
    "Scanrequest",
    "ScanrequestTypedDict",
    "Scanresponse",
    "ScanresponseType",
    "ScanresponseTypedDict",
    "Security",
    "SecurityTypedDict",
    "Textualdetection",
    "TextualdetectionType",
    "TextualdetectionTypedDict",
    "Type",
]
