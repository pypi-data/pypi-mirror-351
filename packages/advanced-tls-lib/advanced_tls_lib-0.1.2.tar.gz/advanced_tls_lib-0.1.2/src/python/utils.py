import re
import json
from typing import Dict, List, Optional, Tuple, Union
from urllib.parse import urlparse, parse_qs
import base64


def parse_headers(header_string: str) -> Dict[str, str]:
    """Parse HTTP headers from string."""
    headers = {}
    lines = header_string.strip().split('\n')
    
    for line in lines:
        if ':' in line:
            key, value = line.split(':', 1)
            headers[key.strip().lower()] = value.strip()
            
    return headers


def build_request_headers(
    browser_chars,
    hostname: str,
    method: str = 'GET',
    path: str = '/',
    additional_headers: Optional[Dict[str, str]] = None
) -> Dict[str, str]:
    """Build complete request headers for browser simulation."""
    headers = {
        'Host': hostname,
        'User-Agent': browser_chars.user_agent,
        'Accept': browser_chars.accept_headers['Accept'],
        'Accept-Language': browser_chars.accept_headers['Accept-Language'],
        'Accept-Encoding': browser_chars.accept_headers['Accept-Encoding'],
        'Connection': 'keep-alive',
    }
    
    # Add Sec-CH-UA headers if available
    if hasattr(browser_chars, 'sec_ch_ua') and browser_chars.sec_ch_ua:
        for key, value in browser_chars.sec_ch_ua.items():
            headers[key] = value
            
    # Add method-specific headers
    if method == 'GET':
        headers['Upgrade-Insecure-Requests'] = '1'
        
    # Add additional headers if provided
    if additional_headers:
        headers.update(additional_headers)
        
    return headers


def parse_response(raw_response: str, request_url: str):
    """Parse raw HTTP response."""
    from .client import Response
    
    # Split response into headers and body
    if '\r\n\r\n' in raw_response:
        header_part, body_part = raw_response.split('\r\n\r\n', 1)
    elif '\n\n' in raw_response:
        header_part, body_part = raw_response.split('\n\n', 1)
    else:
        header_part = raw_response
        body_part = ''
        
    # Parse status line
    lines = header_part.split('\n')
    status_line = lines[0]
    
    # Extract status code
    status_match = re.match(r'HTTP/[\d.]+ (\d+)', status_line)
    status_code = int(status_match.group(1)) if status_match else 200
    
    # Parse headers
    headers = {}
    for line in lines[1:]:
        if ':' in line:
            key, value = line.split(':', 1)
            headers[key.strip().lower()] = value.strip()
            
    # Determine encoding
    encoding = 'utf-8'
    content_type = headers.get('content-type', '')
    if 'charset=' in content_type:
        encoding = content_type.split('charset=')[-1].strip()
        
    return Response(
        status_code=status_code,
        headers=headers,
        body=body_part.encode(encoding),
        url=request_url,
        encoding=encoding
    )


def encode_form_data(data: Dict[str, str]) -> str:
    """Encode form data for POST requests."""
    from urllib.parse import urlencode
    return urlencode(data)


def encode_json_data(data: Dict) -> str:
    """Encode JSON data for POST requests."""
    return json.dumps(data, separators=(',', ':'))


def decode_response_body(body: bytes, encoding: str = 'utf-8') -> str:
    """Decode response body with proper encoding."""
    try:
        return body.decode(encoding)
    except UnicodeDecodeError:
        # Try common encodings
        for enc in ['utf-8', 'latin-1', 'cp1252']:
            try:
                return body.decode(enc)
            except UnicodeDecodeError:
                continue
        # Fallback
        return body.decode('utf-8', errors='replace')


def extract_cookies(headers: Dict[str, str]) -> Dict[str, str]:
    """Extract cookies from Set-Cookie headers."""
    cookies = {}
    set_cookie = headers.get('set-cookie', '')
    
    if isinstance(set_cookie, list):
        cookie_strings = set_cookie
    else:
        cookie_strings = [set_cookie] if set_cookie else []
        
    for cookie_string in cookie_strings:
        # Simple cookie parsing
        parts = cookie_string.split(';')[0].split('=', 1)
        if len(parts) == 2:
            key, value = parts
            cookies[key.strip()] = value.strip()
            
    return cookies


def build_cookie_header(cookies: Dict[str, str]) -> str:
    """Build Cookie header from cookie dict."""
    return '; '.join(f'{k}={v}' for k, v in cookies.items())


def validate_url(url: str) -> bool:
    """Validate URL format."""
    try:
        parsed = urlparse(url)
        return bool(parsed.scheme and parsed.netloc)
    except:
        return False


def extract_domain(url: str) -> str:
    """Extract domain from URL."""
    parsed = urlparse(url)
    return parsed.netloc


def is_redirect_status(status_code: int) -> bool:
    """Check if status code is a redirect."""
    return 300 <= status_code < 400


def get_redirect_url(response_headers: Dict[str, str], base_url: str) -> Optional[str]:
    """Get redirect URL from response headers."""
    location = response_headers.get('location')
    if not location:
        return None
        
    # Handle relative URLs
    if location.startswith('/'):
        parsed_base = urlparse(base_url)
        return f"{parsed_base.scheme}://{parsed_base.netloc}{location}"
    elif not location.startswith('http'):
        return f"{base_url.rstrip('/')}/{location.lstrip('/')}"
    else:
        return location


def calculate_content_length(data: Union[str, bytes]) -> int:
    """Calculate content length for request body."""
    if isinstance(data, str):
        return len(data.encode('utf-8'))
    return len(data)


def generate_boundary() -> str:
    """Generate boundary for multipart/form-data."""
    import random
    import string
    chars = string.ascii_letters + string.digits
    return ''.join(random.choices(chars, k=16))


def encode_multipart_form_data(
    fields: Dict[str, str],
    files: Dict[str, Tuple[str, bytes, str]] = None
) -> Tuple[str, bytes]:
    """Encode multipart/form-data."""
    boundary = generate_boundary()
    content_type = f'multipart/form-data; boundary={boundary}'
    
    body_parts = []
    
    # Add form fields
    for key, value in fields.items():
        part = f'--{boundary}\r\n'
        part += f'Content-Disposition: form-data; name="{key}"\r\n\r\n'
        part += f'{value}\r\n'
        body_parts.append(part.encode('utf-8'))
        
    # Add files
    if files:
        for key, (filename, data, content_type_file) in files.items():
            part = f'--{boundary}\r\n'
            part += f'Content-Disposition: form-data; name="{key}"; filename="{filename}"\r\n'
            part += f'Content-Type: {content_type_file}\r\n\r\n'
            body_parts.append(part.encode('utf-8'))
            body_parts.append(data)
            body_parts.append(b'\r\n')
            
    # Add closing boundary
    body_parts.append(f'--{boundary}--\r\n'.encode('utf-8'))
    
    body = b''.join(body_parts)
    return content_type, body


def is_json_response(headers: Dict[str, str]) -> bool:
    """Check if response is JSON."""
    content_type = headers.get('content-type', '')
    return 'application/json' in content_type


def is_html_response(headers: Dict[str, str]) -> bool:
    """Check if response is HTML."""
    content_type = headers.get('content-type', '')
    return 'text/html' in content_type


def extract_charset(content_type: str) -> str:
    """Extract charset from Content-Type header."""
    if 'charset=' in content_type:
        charset = content_type.split('charset=')[-1].strip()
        return charset.split(';')[0].strip()
    return 'utf-8'


def normalize_header_name(name: str) -> str:
    """Normalize header name to lowercase."""
    return name.lower().strip()


def normalize_header_value(value: str) -> str:
    """Normalize header value."""
    return value.strip()


def parse_quality_values(header_value: str) -> List[Tuple[str, float]]:
    """Parse quality values from Accept headers."""
    items = []
    for item in header_value.split(','):
        item = item.strip()
        if ';q=' in item:
            value, quality = item.split(';q=', 1)
            try:
                quality = float(quality)
            except ValueError:
                quality = 1.0
        else:
            value = item
            quality = 1.0
        items.append((value.strip(), quality))
    
    return sorted(items, key=lambda x: x[1], reverse=True)


def build_user_agent(browser: str, version: str, os: str) -> str:
    """Build User-Agent string."""
    templates = {
        'chrome': 'Mozilla/5.0 ({os}) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/{version} Safari/537.36',
        'firefox': 'Mozilla/5.0 ({os}) Gecko/20100101 Firefox/{version}',
        'safari': 'Mozilla/5.0 ({os}) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/{version} Safari/605.1.15',
    }
    
    template = templates.get(browser.lower(), templates['chrome'])
    return template.format(os=os, version=version)


def escape_header_value(value: str) -> str:
    """Escape header value for HTTP transmission."""
    # Remove control characters
    value = re.sub(r'[\x00-\x1F\x7F]', '', value)
    return value


def detect_encoding(body: bytes, headers: Dict[str, str]) -> str:
    """Detect encoding from headers and content."""
    # Check Content-Type header
    content_type = headers.get('content-type', '')
    if 'charset=' in content_type:
        return extract_charset(content_type)
        
    # Check HTML meta tag
    try:
        html_start = body[:2048].decode('latin-1')
        charset_match = re.search(
            r'<meta[^>]+charset[="\s]+([^">\s]+)',
            html_start,
            re.IGNORECASE
        )
        if charset_match:
            return charset_match.group(1).lower()
    except:
        pass
        
    # Default to UTF-8
    return 'utf-8'


def sanitize_filename(filename: str) -> str:
    """Sanitize filename for safe file operations."""
    # Remove dangerous characters
    filename = re.sub(r'[<>:"/\\|?*]', '_', filename)
    filename = filename.strip('. ')
    
    # Ensure not empty
    if not filename:
        filename = 'unnamed'
        
    return filename