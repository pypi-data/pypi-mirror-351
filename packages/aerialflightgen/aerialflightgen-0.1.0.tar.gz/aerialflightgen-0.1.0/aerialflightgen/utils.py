# Utility functions for coordinate math or file IO

def meters_to_time(distance_m: float, speed_mps: float) -> float:
    return round(distance_m / speed_mps, 2)

def print_banner():
    print("Aerial Flight Generator is active.")
