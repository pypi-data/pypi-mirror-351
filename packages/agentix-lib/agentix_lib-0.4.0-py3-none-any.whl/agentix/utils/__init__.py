from .debug_logger import DebugLogger

__all__ = ["DebugLogger"] 