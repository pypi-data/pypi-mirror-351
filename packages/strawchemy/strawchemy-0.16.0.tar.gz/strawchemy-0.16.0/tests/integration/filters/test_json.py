from __future__ import annotations

from typing import TYPE_CHECKING, Any

import pytest

from sqlalchemy import Insert, MetaData, insert
from syrupy.assertion import SnapshotAssertion
from tests.integration.models import JSONModel, json_metadata
from tests.integration.types import mysql as mysql_types
from tests.integration.types import postgres as postgres_types
from tests.integration.utils import to_graphql_representation
from tests.utils import maybe_async

if TYPE_CHECKING:
    from strawchemy.typing import SupportedDialect

    from tests.integration.fixtures import QueryTracker
    from tests.integration.typing import RawRecordData
    from tests.typing import AnyQueryExecutor


@pytest.fixture
def metadata() -> MetaData:
    return json_metadata


@pytest.fixture
def seed_insert_statements(raw_json: RawRecordData) -> list[Insert]:
    return [insert(JSONModel).values(raw_json)]


@pytest.fixture
def async_query(dialect: SupportedDialect) -> type[Any]:
    if dialect == "postgresql":
        return postgres_types.JSONAsyncQuery
    if dialect == "mysql":
        return mysql_types.JSONAsyncQuery
    pytest.skip(f"JSON tests can't be run on this dialect: {dialect}")


@pytest.fixture
def sync_query(dialect: SupportedDialect) -> type[Any]:
    if dialect == "postgresql":
        return postgres_types.JSONSyncQuery
    if dialect == "mysql":
        return mysql_types.JSONSyncQuery
    pytest.skip(f"JSON tests can't be run on this dialect: {dialect}")


# Tests for JSON-specific filters
@pytest.mark.parametrize(
    ("filter_name", "value", "expected_ids"),
    [
        pytest.param("contains", {"key1": "value1"}, [0], id="contains"),
        pytest.param(
            "containedIn",
            {"key1": "value1", "key2": 2, "nested": {"inner": "value"}, "extra": "value"},
            [0, 2],
            id="containedIn",
        ),
        pytest.param("hasKey", "key1", [0], id="hasKey"),
        pytest.param("hasKeyAll", ["key1", "key2"], [0], id="hasKeyAll"),
        pytest.param("hasKeyAny", ["key1", "status"], [0, 1], id="hasKeyAny"),
    ],
)
@pytest.mark.snapshot
async def test_json_filters(
    filter_name: str,
    value: Any,
    expected_ids: list[int],
    any_query: AnyQueryExecutor,
    raw_json: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    if isinstance(value, list):
        value_str = ", ".join(to_graphql_representation(v, "input") for v in value)
        value_repr = f"[{value_str}]"
    else:
        value_repr = to_graphql_representation(value, "input")

    query = f"""
        {{
            json(filter: {{ dictCol: {{ {filter_name}: {value_repr} }} }}) {{
                id
                dictCol
            }}
        }}
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert len(result.data["json"]) == len(expected_ids)
    for i, expected_id in enumerate(expected_ids):
        assert result.data["json"][i]["id"] == raw_json[expected_id]["id"]
    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot
