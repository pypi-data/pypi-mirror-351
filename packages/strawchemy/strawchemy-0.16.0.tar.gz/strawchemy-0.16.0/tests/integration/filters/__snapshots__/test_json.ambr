# serializer version: 1
# name: test_json_filters[session-tracked-async-containedIn-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_contains(CAST(%s AS JSON), json_model.dict_col)
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-containedIn-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col <@ $1::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-containedIn-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col <@ %(param_1)s::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-contains-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_contains(json_model.dict_col, CAST(%s AS JSON))
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-contains-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col @> $1::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-contains-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col @> %(param_1)s::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKey-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_contains_path(json_model.dict_col, %s, %s)
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKey-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? $1::VARCHAR
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKey-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? %(param_1)s::VARCHAR
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAll-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_contains_path(json_model.dict_col, %s, %s, %s)
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAll-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? & CAST($1::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAll-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? & CAST(%(param_1)s::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAny-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_contains_path(json_model.dict_col, %s, %s, %s)
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAny-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? | CAST($1::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAny-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? | CAST(%(param_1)s::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-containedIn-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col <@ %(param_1)s::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-contains-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col @> %(param_1)s::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-hasKey-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? %(param_1)s::VARCHAR
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-hasKeyAll-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? & CAST(%(param_1)s::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-hasKeyAny-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? | CAST(%(param_1)s::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
